### PR TITLE
2408.3 - refactor tanf t1 parsing

### LINKED ISSUE #2408: Preparser Refactor
**Description:**
As discussed during Friday 2/10 with the dev team, this ticket is scoped for refactoring and abstracting the head of parsing.

**Acceptance Criteria:** 
- [ ] Parsing is a standalone django app
- [ ] Preparsing and overall parsing management are abstracted from each other
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
- [x] Move parsing to its own app and reorganize folder structure (`tdpservice/search_indexes/parsers` --> `tdpservice/parsers`)
- [x] Move preparsing error checks (cat1) into a separate file/set of functions
- [x] Refactor existing `preparse()` to be an overall parsing manager/main function and subordinate `preparse` function
- [ ] STRETCH: Preparsing record types header/trailer are created as models akin to T1 documents
- [ ] STRETCH: All category 1 errors utilize data struct/model from #1121 
- [x] STRETCH: Adjust parsing unit tests for restructuring in this ticket
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
These notes are meant to be a guideline and to flesh out the ideas above but are not definitive or immovable.
* Create new file `parsers/parser_manager.py` to be invoked by `scheduling.parser_task.py`.
* This new manager will call parsing functions in order based on DataFile object (`data_type`, `section`, etc.)
```python
# tdpservice/parsers/manager.py
def main(datafile, data_type, section):
  # datafile handling
  returns = preparse(kwargs)
  # error-handling of preparsing
  
    if data_type == 'TANF':
        if section == "Active Case Data":
           tanf_parser.active_parse(datafile)
       elif section == "Closed Case Data":
          tanf_parser.closed_parse(datafile)
    #   we don't currently have different sections but we'll stub those out here
    
    # elif data_type == 'SSP':
    #    ssp_parser.parse(datafile)
    # elif data_type == 'Tribal TANF':
    #    tribal_tanf_parser.parse(datafile)
    else:
        raise ValueError('Preparser given invalid data_type parameter.')
```
* Any sort of structure errors or cat-1 validation should be left in `preparser.py`.
* Per supporting documentation, we will treat preparsing like a "section 0" to be done before section 1. This includes refactoring header/trailer validation to match structure of #1121
* Any category 1 checks on data rows (e.g., length checks, bad record type) should be included as part of preparsing.
* Header content must be saved in a new model so it can be referenced in later validation checks.
* This will require significant reworking of unit tests
* Another note, per repeated commentary from @ADPennington, we should treat any trailer mistakes as warnings and just proceed with all relevant parsing.

**Supporting Documentation:**
- [Flow Diagram](https://app.mural.co/t/raft2792/m/raft2792/1675268418098/d56ec0d5415d8cda9c6ae9cae61154bcc4e74bba?sender=ce75cff7-c7f8-41b1-8947-5b7df76ae538)
- [Detailed Architecture Description](https://lucid.app/lucidchart/a86a36e7-dd40-4d10-bff9-d5523212f322/edit?invitationId=inv_d6e2304a-12f8-49fb-adc8-20e2d640adc8&page=0_0#)

**Open Questions:**

### PR DESCRIPTION
**Pull request by jtimpe**
## Summary of Changes
Pull request closes #2408  

* Refactor implementation of `tanf.T1` to utilize the new `RowSchema` utility class.
* Implement the new schema in the `parse_datafile` function
* Update automated tests

This PR is an example of how to implement future parsing schemas within the new structure.

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && npm run start
cd tdrs-backend && docker-compose up
```

1. Open http://localhost:3000/ and sign in.
2. Grant your user an "Approved" account status and assign them to a "Data Analyst" role.
3. Navigate to the Data Files page and upload `small_correct_file` from `tdpservice/parsing/test/data` to "Section 1" for any quarter and any year.
4. Navigate to the django admin and view the newly created "T1" document. Make sure field values match expected values.

This PR does not introduce validation of the T1 document, which will be implemented with #2369.

## Deliverables
_More details on how deliverables herein are assessed included [here](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [ ] [**_insert ACs here_**]
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/parsers/schema_defs/__init__.py [hunk 1]
```diff
@@ -1,5 +1,7 @@
 from .header import header
 from .trailer import trailer
+from . import tanf
 
 header = header
 trailer = trailer
+tanf = tanf
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/parsers/schema_defs/__init__.py:7]: Why do we need these assignments?

Comment (jtimpe) [reply tdrs-backend/tdpservice/parsers/schema_defs/__init__.py:7]: without them, we'd have to `import t1 from tdpservice.parsers.schema_defs.tanf.t1`. with it we only have to `import t1 from tdpservice.parsers.schema_defs.tanf` - hope that makes sense. it just normalizes the imports a little, similar to `export` in javascript modules

Comment (raftmsohani) [reply tdrs-backend/tdpservice/parsers/schema_defs/__init__.py:7]: Nit: I would vote against it and would remove them and go with the second approach for importing.

**Review by raftmsohani — APPROVED** [review]
LGTM!

**Review by andrew-jameson — APPROVED** [review]
Looking good, I think we're set up nicely for tanf section 1 validation here now. Also, running through some file submissions, we might need a more concise logging mechanism or compartmentalizing logs for specific files -- routing those to separate log files, food for thought on this.

**Review by ADPennington — APPROVED** [review]
lgtm @jtimpe thanks! I tried a few different iterations to review the logs:

- small clean file (TANF, section 1, T1 only) :heavy_check_mark:

```
 2023-04-07T12:47:33.46-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:47:33,466 DEBUG validators.py::validate_file_infection:L54 :  CLAMAV_NEEDED noted as True, proceeding with scan.
2023-04-07T12:47:33.46-0400 [APP/PROC/WEB/0] ERR CLAMAV_NEEDED noted as True, proceeding with scan.
2023-04-07T12:47:33.46-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:47:33,466 DEBUG clients.py::__init__:L37 :  Set clamav endpoint_url as 'http://tanf-dev-clamav-rest.apps.internal:9000/scan'
2023-04-07T12:47:33.46-0400 [APP/PROC/WEB/0] ERR Set clamav endpoint_url as 'http://tanf-dev-clamav-rest.apps.internal:9000/scan'
2023-04-07T12:47:33.46-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:47:33,466 DEBUG clients.py::scan_file:L64 :  Initiating virus scan for file: small_correct_file_T1.txt
2023-04-07T12:47:33.46-0400 [APP/PROC/WEB/0] ERR Initiating virus scan for file: small_correct_file_T1.txt
2023-04-07T12:47:33.48-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:47:33,479 DEBUG clients.py::scan_file:L90 :  File scan marked as CLEAN for file: small_correct_file_T1.txt
2023-04-07T12:47:33.48-0400 [APP/PROC/WEB/0] ERR File scan marked as CLEAN for file: small_correct_file_T1.txt
2023-04-07T12:47:33.57-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:47:33,577 INFO views.py::create:L71 :  Submitted parse task to queue for datafile 340.
2023-04-07T12:47:33.57-0400 [APP/PROC/WEB/0] ERR Submitted parse task to queue for datafile 340.
2023-04-07T12:47:33.57-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:47:33,578 INFO views.py::create:L80 :  Submitted upload task to redis for datafile 340.
2023-04-07T12:47:33.57-0400 [APP/PROC/WEB/0] ERR Submitted upload task to redis for datafile 340.
2023-04-07T12:47:33.59-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:47:33,595 INFO parser_task.py::parse:L19 :  DataFile parsing started for file ADS.E2J.NDM1.TS49
2023-04-07T12:47:33.59-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:47:33,595: INFO/ForkPoolWorker-1] DataFile parsing started for file ADS.E2J.NDM1.TS49
2023-04-07T12:47:33.80-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:47:33,803 INFO parser_task.py::parse:L21 :  DataFile parsing finished with 0 errors: {}
2023-04-07T12:47:33.80-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:47:33,803: INFO/ForkPoolWorker-1] DataFile parsing finished with 0 errors: {}
2023-04-07T12:47:36.12-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:47:36,122 INFO sftp_task.py::upload:L97 :  File ADS.E2J.NDM1.TS49 has been successfully uploaded to acftitan.cit.nih.gov
2023-04-07T12:47:36.12-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:47:36,122: INFO/ForkPoolWorker-1] File ADS.E2J.NDM1.TS49 has been successfully uploaded to acftitan.cit.nih.gov
   
```

- small, clean file (TANF, Section 2, T1 only) :heavy_check_mark:

```
2023-04-07T12:51:00.18-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:51:00 +0000] [568] [DEBUG] POST /v1/data_files/
   2023-04-07T12:51:00.21-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:51:00,210 DEBUG validators.py::validate_file_infection:L54 :  CLAMAV_NEEDED noted as True, proceeding with scan.
   2023-04-07T12:51:00.21-0400 [APP/PROC/WEB/0] ERR CLAMAV_NEEDED noted as True, proceeding with scan.
   2023-04-07T12:51:00.21-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:51:00,210 DEBUG clients.py::__init__:L37 :  Set clamav endpoint_url as 'http://tanf-dev-clamav-rest.apps.internal:9000/scan'
   2023-04-07T12:51:00.21-0400 [APP/PROC/WEB/0] ERR Set clamav endpoint_url as 'http://tanf-dev-clamav-rest.apps.internal:9000/scan'
   2023-04-07T12:51:00.21-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:51:00,210 DEBUG clients.py::scan_file:L64 :  Initiating virus scan for file: small_correct_file_T1.txt
   2023-04-07T12:51:00.21-0400 [APP/PROC/WEB/0] ERR Initiating virus scan for file: small_correct_file_T1.txt
   2023-04-07T12:51:00.21-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:51:00,217 DEBUG clients.py::scan_file:L90 :  File scan marked as CLEAN for file: small_correct_file_T1.txt
   2023-04-07T12:51:00.21-0400 [APP/PROC/WEB/0] ERR File scan marked as CLEAN for file: small_correct_file_T1.txt
   2023-04-07T12:51:00.30-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:51:00,302 INFO views.py::create:L71 :  Submitted parse task to queue for datafile 341.
   2023-04-07T12:51:00.30-0400 [APP/PROC/WEB/0] ERR Submitted parse task to queue for datafile 341.
   2023-04-07T12:51:00.30-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:51:00,303 INFO views.py::create:L80 :  Submitted upload task to redis for datafile 341.
   2023-04-07T12:51:00.30-0400 [APP/PROC/WEB/0] ERR Submitted upload task to redis for datafile 341.
   2023-04-07T12:51:00.31-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:51:00,319 INFO parser_task.py::parse:L19 :  DataFile parsing started for file ADS.E2J.NDM2.TS49
   2023-04-07T12:51:00.31-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:51:00,319: INFO/ForkPoolWorker-1] DataFile parsing started for file ADS.E2J.NDM2.TS49
   2023-04-07T12:51:00.41-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:51:00,410 INFO parser_task.py::parse:L21 :  DataFile parsing finished with 1 errors: {'document': ['Section does not match.']}
   2023-04-07T12:51:00.41-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:51:00,410: INFO/ForkPoolWorker-1] DataFile parsing finished with 1 errors: {'document': ['Section does not match.']}
   2023-04-07T12:51:00.49-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:51:00 +0000] [568] [DEBUG] POST /v1/logs/
   2023-04-07T12:51:02.32-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:51:02,321 INFO sftp_task.py::upload:L97 :  File ADS.E2J.NDM2.TS49 has been successfully uploaded to acftitan.cit.nih.gov
   2023-04-07T12:51:02.32-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:51:02,321: INFO/ForkPoolWorker-1] File ADS.E2J.NDM2.TS49 has been successfully uploaded to acftitan.cit.nih.gov

```

- small not so clean file (TANF, T1+T2+T3, where some T1s had incorrect length) :heavy_check_mark:

```
 2023-04-07T12:55:25.18-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:55:25,182 DEBUG validators.py::validate_file_infection:L54 :  CLAMAV_NEEDED noted as True, proceeding with scan.
   2023-04-07T12:55:25.18-0400 [APP/PROC/WEB/0] ERR CLAMAV_NEEDED noted as True, proceeding with scan.
   2023-04-07T12:55:25.18-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:55:25,182 DEBUG clients.py::__init__:L37 :  Set clamav endpoint_url as 'http://tanf-dev-clamav-rest.apps.internal:9000/scan'
   2023-04-07T12:55:25.18-0400 [APP/PROC/WEB/0] ERR Set clamav endpoint_url as 'http://tanf-dev-clamav-rest.apps.internal:9000/scan'
   2023-04-07T12:55:25.18-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:55:25,183 DEBUG clients.py::scan_file:L64 :  Initiating virus scan for file: small_notclean_file_T1.txt
   2023-04-07T12:55:25.18-0400 [APP/PROC/WEB/0] ERR Initiating virus scan for file: small_notclean_file_T1.txt
   2023-04-07T12:55:25.19-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:55:25,196 DEBUG clients.py::scan_file:L90 :  File scan marked as CLEAN for file: small_notclean_file_T1.txt
   2023-04-07T12:55:25.19-0400 [APP/PROC/WEB/0] ERR File scan marked as CLEAN for file: small_notclean_file_T1.txt
   2023-04-07T12:55:25.33-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:55:25,333 INFO views.py::create:L71 :  Submitted parse task to queue for datafile 342.
   2023-04-07T12:55:25.33-0400 [APP/PROC/WEB/0] ERR Submitted parse task to queue for datafile 342.
   2023-04-07T12:55:25.33-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:55:25,334 INFO views.py::create:L80 :  Submitted upload task to redis for datafile 342.
   2023-04-07T12:55:25.33-0400 [APP/PROC/WEB/0] ERR Submitted upload task to redis for datafile 342.
 2023-04-07T12:55:25.35-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:55:25,350 INFO parser_task.py::parse:L19 :  DataFile parsing started for file ADS.E2J.NDM1.TS49
   2023-04-07T12:55:25.35-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:55:25,350: INFO/ForkPoolWorker-1] DataFile parsing started for file ADS.E2J.NDM1.TS49
   2023-04-07T12:55:25.40-0400 [APP/PROC/WEB/0] OUT 10.255.230.198 - - [07/Apr/2023:16:55:25 +0000] "POST /v1/data_files/ HTTP/1.1" 201 301 "https://tdp-frontend-qasp.app.cloud.gov/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36"
   2023-04-07T12:55:25.50-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:55:25,502 INFO parser_task.py::parse:L21 :  DataFile parsing finished with 4 errors: {3: ['No schema selected.'], 4: ['No schema selected.'], 5: ['Value length 158 does not match 156.'], 6: ['Value length 154 does not match 156.']}
   2023-04-07T12:55:25.50-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:55:25,502: INFO/ForkPoolWorker-1] DataFile parsing finished with 4 errors: {3: ['No schema selected.'], 4: ['No schema selected.'], 5: ['Value length 158 does not match 156.'], 6: ['Value length 154 does not match 156.']}
   2023-04-07T12:55:25.52-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:55:25 +0000] [567] [DEBUG] POST /v1/logs/
2023-04-07T12:55:27.53-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:55:27,539 INFO sftp_task.py::upload:L97 :  File ADS.E2J.NDM1.TS49 has been successfully uploaded to acftitan.cit.nih.gov
   2023-04-07T12:55:27.53-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:55:27,539: INFO/ForkPoolWorker-1] File ADS.E2J.NDM1.TS49 has been successfully uploaded to acftitan.cit.nih.gov
   
```

- small clean file for SSP (TANF, Section1,  T1) :heavy_check_mark:
```
 2023-04-07T12:59:35.10-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:59:35,103 INFO parser_task.py::parse:L21 :  DataFile parsing finished with 1 errors: {'document': ['Section does not match.']}
   2023-04-07T12:59:35.10-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:59:35,103: INFO/ForkPoolWorker-1] DataFile parsing finished with 1 errors: {'document': ['Section does not match.']}
   2023-04-07T12:59:35.20-0400 [APP/PROC/WEB/0] ERR [2023-04-07 16:59:35 +0000] [568] [DEBUG] POST /v1/logs/
   
```

- small clean file for SSP (TANF, Section 2, T1) :heavy_check_mark:
```
2023-04-07T13:00:41.88-0400 [APP/PROC/WEB/0] ERR [2023-04-07 17:00:41,889 INFO parser_task.py::parse:L21 :  DataFile parsing finished with 1 errors: {'document': ['Section does not match.']}
   2023-04-07T13:00:41.88-0400 [APP/PROC/WEB/0] ERR [2023-04-07 17:00:41,889: INFO/ForkPoolWorker-1] DataFile parsing finished with 1 errors: {'document': ['Section does not match.']}
   2023-04-07T13:00:41.97-0400 [APP/PROC/WEB/0] ERR [2023-04-07 17:00:41 +0000] [568] [DEBUG] POST /v1/logs/
   
```

- small, not clean file for TAN (ssp in header) :heavy_check_mark:
```
 2023-04-07T13:02:11.55-0400 [APP/PROC/WEB/0] ERR [2023-04-07 17:02:11,554 INFO parser_task.py::parse:L21 :  DataFile parsing finished with 1 errors: {'document': ['Section does not match.']}
   2023-04-07T13:02:11.55-0400 [APP/PROC/WEB/0] ERR [2023-04-07 17:02:11,554: INFO/ForkPoolWorker-1] DataFile parsing finished with 1 errors: {'document': ['Section does not match.']}
   2023-04-07T13:02:11.63-0400 [APP/PROC/WEB/0] ERR [2023-04-07 17:02:11 +0000] [569] [DEBUG] POST /v1/logs/
 
```

**Pull request merged**